src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2571,74 +2580,97 @@ namespace QgsWms
   {
-    QString featureInfoString;
-
-    //the HTML head
-    featureInfoString.append( "<HEAD>\n" );
-    featureInfoString.append( "<TITLE> GetFeatureInfo results </TITLE>\n" );
-    featureInfoString.append( "<META http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\"/>\n" );
-    featureInfoString.append( "</HEAD>\n" );
-
-    //start the html body
-    featureInfoString.append( "<BODY>\n" );
+    QString featureInfoString = QStringLiteral( R"HTML(
+    <!DOCTYPE html>
+    <head>
+    <title>Information</title>
+    <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
+    <style>
+     body {
+       font-family: 'Open Sans','Calluna Sans','Gill Sans MT',Calibri,'Trebuchet MS',sans-serif;
+     }
+     table, th, td {
+       width: 100%;
+       border: 1px solid black;
+       border-collapse: collapse;
+       text-align: left;
+       padding: 2px;
+     }
+     th {
+       width: 25%;
+       font-weight: bold;
+     }
+     .layer-title {
+       font-weight: bold;
+       padding: 2px;
+     }
+    </style>
+    </head>
+    <body>)HTML" );
 
-    QDomNodeList layerList = doc.elementsByTagName( QStringLiteral( "Layer" ) );
+    const QDomNodeList layerList = doc.elementsByTagName( QStringLiteral( "Layer" ) );
 
     //layer loop
     for ( int i = 0; i < layerList.size(); ++i )
     {
-      QDomElement layerElem = layerList.at( i ).toElement();
-
-      featureInfoString.append( "<TABLE border=1 width=100%>\n" );
-      featureInfoString.append( "<TR><TH width=25%>Layer</TH><TD>" + layerElem.attribute( QStringLiteral( "name" ) ) + "</TD></TR>\n" );
-      featureInfoString.append( "</BR>" );
+      const QDomElement layerElem = layerList.at( i ).toElement();
 
       //feature loop (for vector layers)
-      QDomNodeList featureNodeList = layerElem.elementsByTagName( QStringLiteral( "Feature" ) );
-      QDomElement currentFeatureElement;
+      const QDomNodeList featureNodeList = layerElem.elementsByTagName( QStringLiteral( "Feature" ) );
+      const QDomElement currentFeatureElement;
 
       if ( !featureNodeList.isEmpty() ) //vector layer
       {
+        const QString featureInfoLayerTitleString = QStringLiteral( R"HTML(<div class='layer-title'>%1</div>)HTML" ).arg( layerElem.attribute( QStringLiteral( "title" ) ) );
+        featureInfoString.append( featureInfoLayerTitleString );
+
         for ( int j = 0; j < featureNodeList.size(); ++j )
         {
-          QDomElement featureElement = featureNodeList.at( j ).toElement();
-          featureInfoString.append( "<TABLE border=1 width=100%>\n" );
-          featureInfoString.append( "<TR><TH>Feature</TH><TD>" + featureElement.attribute( QStringLiteral( "id" ) ) +
-                                    "</TD></TR>\n" );
+          const QDomElement featureElement = featureNodeList.at( j ).toElement();
+          featureInfoString.append( QStringLiteral( R"HTML(<table>)HTML" ) );
```

Comment:
Why are you using 'R'aw strings? There aren't any quotes.

Reply:
They were included in the original commit and I didn't change them, so I can only guess that they are used because there are quotes in the head<br>
-> https://github.com/qgis/QGIS/pull/55751/files#diff-2b31f4f76dacb4bd2240768e67fbe59d416219b9b58b3fece7706a1ebb18fbaeR2587)<br>
and also to get a nice CSS formatting without a need to use `\n`.
 
That said, if you don't want to use 'R'aw strings I'm happy to replace them.

Reply:
I am a big fan of raw strings and I wasn't complaining about raw strings when they are useful (like https://github.com/qgis/QGIS/pull/55751/files#diff-2b31f4f76dacb4bd2240768e67fbe59d416219b9b58b3fece7706a1ebb18fbaeR2583) but when they aren't, so please remove them unless necessary/useful (i.e to avoid backslash quoting of double quotes or newlines).

Reply:
Done in https://github.com/qgis/QGIS/pull/55751/commits/b9f2825204b544908641b5cbc16e595c86422344 along with HTML formatting improvements.

Reply:
> Done in [b9f2825](https://github.com/qgis/QGIS/commit/b9f2825204b544908641b5cbc16e595c86422344) along with HTML formatting improvements.

Looks good!

I have just a final request: we should make sure that the code behaves correctly when HTML tag chars are used in the strings, can you please add test cases with:

1. layer name containing `<` and `>`
2. layer title containing `<` and `>`
3. attribute name  containing `<` and `>`
4. attribute value  containing `<` and `>`

please check that valid HTML is produced.

Reply:
Good point :+1: , this was an issue also before this PR :wink:

Addressed per https://github.com/qgis/QGIS/pull/55751/commits/fddb2363989f088baf79c237aefa4a27dc3083f1

I've used [toHtmlEscaped()](https://doc.qt.io/qt-5/qstring.html#toHtmlEscaped) except for the maptip attribute (if present) because HTML could be wanted here.

I think it's the responsibility of the user to provide valid content for the maptip.

Reply:
> Good point 👍 , this was an issue also before this PR 😉

Yeah, I know it isn't your fault, thank you for fixing this!

> 
> Addressed per [fddb236](https://github.com/qgis/QGIS/commit/fddb2363989f088baf79c237aefa4a27dc3083f1)
> 
> I've used [toHtmlEscaped()](https://doc.qt.io/qt-5/qstring.html#toHtmlEscaped) except for the maptip attribute (if present) because HTML could be wanted here.
> 
> I think it's the responsibility of the user to provide valid content for the maptip.

Yes, that makes sense.

Reply:
Thank you too.